File: lintcode/python/0609_two_sum_less_than_or_equal_to_target.py
#!/usr/bin/python -t

# two pointers

class Solution:
    """
    @param nums: an array of integer
    @param target: an integer
    @return: an integer
    """
    def twoSum5(self, nums, target):
        # write your code here
        n = len(nums)
        ret = 0
        nums.sort()
        l = 0
        r = n-1
        
        while l < r:
            val = nums[l] + nums[r]
            if val <= target:
                ret += r-l
                l += 1
            else:
                r -= 1
                
        return ret
                

File: lintcode/golang/0609_two_sum_less_than_or_equal_to_target.go
/**
 * @param nums: an array of integer
 * @param target: an integer
 * @return: an integer
 */

import "sort"

func twoSum5 (nums []int, target int) int {
    // write your code here
    sort.Ints(nums)
    n := len(nums)
    if n == 0 {
        return 0
    }
    
    ret := 0
    l := 0
    r := n-1
    
    for l < r {
        val := nums[l]+nums[r]
        if val <= target {
            ret += r-l
            l++
        } else {
            r--
        }
    }
    
    return ret
}

